src/Stage/Desugar.elm

```diff
@@ -183,12 +189,28 @@ desugarExpr modules thisModule located =
             bindings
                 |> List.map (Binding.map recurse >> Binding.combine)
                 |> Result.combine
-                |> map
+                |> andThen
                     (\canonicalBindings ->
                         canonicalBindings
-                            |> List.map (\canonicalBinding -> ( canonicalBinding.name, canonicalBinding ))
-                            |> Dict.fromList
-                            |> Canonical.Record
+                            |> List.foldr
+                                (\canonicalBinding ->
+                                    Result.andThen
+                                        (\dict ->
+                                            case Dict.get canonicalBinding.name dict of
+                                                Just existing ->
+                                                    Err <|
+                                                        DuplicateRecordField
+                                                            { name = canonicalBinding.name
+                                                            , firstOccurrence = Located.replaceWith () existing.body
+                                                            , secondOccurrence = Located.replaceWith () canonicalBinding.body
+                                                            }
+
```

Comment:
Is it worth generalizing this to something like `findDups .name .body (Located.replaceWith ()) DuplicateRecordField lst`?

Reply:
Honestly, I don't think so. I don't think it's usual at all to have more than 2 duplicate fields.

Reply:
Not sure I understand your comment.  I'm not saying that we should detect more than 2 duplicates.  Do you mean "not sure it's common to do duplicate detection anywhere else in the codebase"?

Reply:
My suggestion may have been unclear--I was thinking just have a helper that detects the first two dups generically.  But I'm not sure we need it yet.  The only other place that seems to do something like duplicate detection does something slightly different (ambiguous module names).

I feel like @Janiczek has some fancy module that collapses a dict into a dict of k -> dups.

Reply:
Something like Dict.Extra.groupBy will surely exist. I wouldn't generalize this yet; we can look at the similar usecases later and see if they're common enough to be generalized. (in another PR possibly)

Reply:
I propose `ensureUnique`: 

``` elm
 65         bindings =
 66             [ { name = "fred", body = "line 3" }
 67             , { name = "mary", body = "line 5" }
 68             , { name = "fred", body = "line 7" }
 69             ]
 70
 71         ensureUnique : List ( comparable, v ) -> Maybe ( comparable, v, v )
 72         ensureUnique list =
 73             let
 74                 uniqueHelper : List ( comparable, v ) -> Maybe ( comparable, v, v )
 75                 uniqueHelper lst =
 76                     case lst of
 77                         [] ->
 78                             Nothing
 79
 80                         [ x ] ->
 81                             Nothing
 82
 83                         ( k1, v1 ) :: ( k2, v2 ) :: rest ->
 84                             if k1 == k2 then
 85                                 Just ( k1, v1, v2 )
 86
 87                             else
 88                                 ( k2, v2 ) :: rest |> uniqueHelper
 89             in
 90             list
 91                 |> List.sortBy Tuple.first
 92                 |> uniqueHelper
 93
 94         uncheckedBindings =
 95             bindings
 96                 |> List.map (\r -> ( r.name, r.body ))
 97
 98         result =
 99             case ensureUnique uncheckedBindings of
100                 Nothing ->
101                     Ok (Dict.fromList uncheckedBindings)
102
103                 Just ( name, body1, body2 ) ->
104                     Err { name = name, body1 = body1, body2 = body2 }
```

Reply:
I'm on mobile but here's what I had in mind
```elm
bindings : List (Binding expr)

bindings
|> Dict.Extra.groupBy .name
|> Dict.Extra.find (\_ occurences -> List.length occurrences > 1)
|> Maybe.map (\name occurrences -> ... the error ...)
|> Maybe.withDefault ...
```

Reply:
Yeah, that seems the way to go!  I think it will make the code a bit easier to read, and it would also be easier to extend to some kind of more verbose error message if somebody had a bunch of duplication for some reason.

Reply:
I went for
```elm
findDuplicatesBy : (a -> comparable) -> List a -> Maybe ( a, a )
findDuplicatesBy property list =
    case list of
        [] ->
            Nothing

        head :: tail ->
            case List.Extra.find (\item -> property item == property head) tail of
                Just duplicate ->
                    Just ( head, duplicate )

                Nothing ->
                    findDuplicatesBy property tail
```

Which should be pretty fast and avoid allocations.